vllm/model_executor/models/qwen2_audio.py

```diff
@@ -0,0 +1,447 @@
+                                               padding="max_length")
+
+    feature_attention_mask = audio_inputs.pop("attention_mask")
+    input_features = audio_inputs['input_features']
+
+    audio_feat_lengths, audio_output_lengths = _get_feat_extract_output_lengths(
+        feature_attention_mask.sum(-1))
+
+    audio_token_index = ctx.model_config.hf_config.audio_token_index
+
+    input_ids = llm_inputs['prompt_token_ids']
+
+    new_input_ids = []
+    audio_num = input_ids.count(audio_token_index)
+    assert len(input_features) == audio_num, \
+        (f'The text input contains {audio_num} audio tokens, '
+         f'but {len(input_features)} audios provided')
+    start = 0
+    for audio_idx in range(audio_num):
+        end = input_ids.index(audio_token_index, start)
+        new_input_ids.extend(input_ids[start:end])  # text part
+
+        new_input_ids.extend([audio_token_index] *
+                             audio_output_lengths[audio_idx])
+        start = end + 1
+    new_input_ids.extend(input_ids[start:])
+
+    return LLMInputs(
+        prompt_token_ids=new_input_ids,
+        prompt=llm_inputs['prompt'],
+        multi_modal_data=multi_modal_data,
+    )
+
+
+def input_mapper_for_qwen2_audio(
+    ctx: InputContext,
+    multi_modal_data: Union[np.ndarray, List[np.ndarray]],
+) -> MultiModalInputs:
+    """Input mapper for Qwen2-Audio."""
+    processor = cached_get_processor(ctx.model_config.model)
+    audio_feature_extractor = processor.feature_extractor
+    if audio_feature_extractor is None:
+        raise RuntimeError(
+            "No HuggingFace audio_feature_extractor is available "
+            "to process the audio object")
+
+    if len(multi_modal_data) == 0:
+        batch_data = {
+            'input_features': torch.tensor([]),
+            'feature_attention_mask': torch.tensor([])
+        }
+        return batch_data
+    try:
+        audios = [_[0] for _ in multi_modal_data]
+        batch_data = audio_feature_extractor(audios,
+                                             sampling_rate=16000,
+                                             return_attention_mask=True,
+                                             padding="max_length",
+                                             return_tensors="pt").data
+        batch_data["feature_attention_mask"] = batch_data.pop("attention_mask")
+    except Exception:
+        logger.error("Failed to process audio (%s)", audios)
+        raise
+
+    return MultiModalInputs(batch_data)
+
+
+@INPUT_REGISTRY.register_dummy_data(dummy_data_for_qwen2_audio)
+@INPUT_REGISTRY.register_input_processor(input_processor_for_qwen2_audio)
+@MULTIMODAL_REGISTRY.register_input_mapper("audio",
+                                           input_mapper_for_qwen2_audio)
+@MULTIMODAL_REGISTRY.register_max_multimodal_tokens(
+    "audio", get_max_qwen2_audio_audio_tokens)
+class Qwen2AudioForConditionalGeneration(nn.Module, SupportsMultiModal):
```

Comment:
Since you're passing in `intermediate_tensors` to the underlying LM already, please implement the methods required for `SupportsPP`. (You can refer to the other multi-modal models for some examples).

Reply:
Hi @DarkLight1337 , I already added `SupportsPP`, could you please have another look? Thanks.

Reply:
After some tweaks, I managed to get the model to work with PP.